core/node/consensus/src/storage/connection.rs

```diff
@@ -239,17 +250,152 @@ impl<'a> Connection<'a> {
+        ctx: &ctx::Ctx,
+    ) -> ctx::Result<Option<attester::BatchNumber>> {
+        Ok(ctx
+            .wait(self.0.blocks_dal().get_sealed_l1_batch_number())
+            .await?
+            .context("get_sealed_l1_batch_number()")?
+            .map(|nr| attester::BatchNumber(nr.0 as u64)))
+    }
+
+    /// Wrapper for `consensus_dal().get_last_batch_certificate_number()`.
+    pub async fn get_last_batch_certificate_number(
+        &mut self,
+        ctx: &ctx::Ctx,
+    ) -> ctx::Result<Option<attester::BatchNumber>> {
+        Ok(ctx
+            .wait(self.0.consensus_dal().get_last_batch_certificate_number())
+            .await?
+            .context("get_last_batch_certificate_number()")?)
+    }
+
+    /// Wrapper for `consensus_dal().batch_certificate()`.
+    pub async fn batch_certificate(
+        &mut self,
+        ctx: &ctx::Ctx,
+        number: attester::BatchNumber,
+    ) -> ctx::Result<Option<attester::BatchQC>> {
+        Ok(ctx
+            .wait(self.0.consensus_dal().batch_certificate(number))
+            .await?
+            .context("batch_certificate()")?)
+    }
+
+    /// Wrapper for `blocks_dal().get_l2_block_range_of_l1_batch()`.
+    pub async fn get_l2_block_range_of_l1_batch(
+        &mut self,
+        ctx: &ctx::Ctx,
+        number: attester::BatchNumber,
+    ) -> ctx::Result<Option<(validator::BlockNumber, validator::BlockNumber)>> {
+        let number = L1BatchNumber(number.0.try_into().context("number")?);
+
+        let range = ctx
+            .wait(self.0.blocks_dal().get_l2_block_range_of_l1_batch(number))
+            .await?
+            .context("get_l2_block_range_of_l1_batch()")?;
+
+        Ok(range.map(|(min, max)| {
+            let min = validator::BlockNumber(min.0 as u64);
+            let max = validator::BlockNumber(max.0 as u64);
+            (min, max)
+        }))
+    }
+
+    /// Construct the [attester::SyncBatch] for a given batch number.
+    pub async fn get_batch(
+        &mut self,
+        ctx: &ctx::Ctx,
+        number: attester::BatchNumber,
+    ) -> ctx::Result<Option<attester::SyncBatch>> {
+        let Some((min, max)) = self
+            .get_l2_block_range_of_l1_batch(ctx, number)
+            .await
+            .context("get_l2_block_range_of_l1_batch()")?
+        else {
+            return Ok(None);
+        };
+
+        let payloads = self.payloads(ctx, min..max).await.wrap("payloads()")?;
+        let payloads = payloads.into_iter().map(|p| p.encode()).collect();
+
+        // TODO: Fill out the proof when we have the stateless L1 batch validation story finished.
+        // It is supposed to be a Merkle proof that the rolling hash of the batch has been included
+        // in the L1 state tree. The state root hash of L1 won't be available in the DB, it requires
+        // an API client.
+        let batch = attester::SyncBatch {
+            number,
+            payloads,
+            proof: Vec::new(),
+        };
+
+        Ok(Some(batch))
```

Comment:
it's not clear for me why we need to load the whole payload for this feature. It should be enough to sign the commitement/root hash - why does consensus side request the whole payload? cc @brunoffranca

Reply:
My understanding from yesterday was that this is the way gossip propagates the batches through the network:
* an RPC server is registered that can [serve batches](https://github.com/matter-labs/era-consensus/blob/decb988eb9e1a45fd5171d2cc540a360d9ca5f1f/node/actors/network/src/gossip/runner.rs#L146) upon request
* this is [called](https://github.com/matter-labs/era-consensus/blob/decb988eb9e1a45fd5171d2cc540a360d9ca5f1f/node/actors/network/src/gossip/runner.rs#L322) whenever the other party indicates it has a new batch
* the `BatchStore` will [look](https://github.com/matter-labs/era-consensus/blob/decb988eb9e1a45fd5171d2cc540a360d9ca5f1f/node/libs/storage/src/batch_store.rs#L231) either in the cache or the database
* the first time `BatchStore` [receives](https://github.com/matter-labs/era-consensus/blob/decb988eb9e1a45fd5171d2cc540a360d9ca5f1f/node/libs/storage/src/batch_store.rs#L111) a batch it will queue it in the cache
* the `BatchStoreRunner` [forwards](https://github.com/matter-labs/era-consensus/blob/decb988eb9e1a45fd5171d2cc540a360d9ca5f1f/node/libs/storage/src/batch_store.rs#L175) anything queued in the cache to the persistent store

So, the payload is for peers that don't have them because they missed all the L2 blocks. Now they get them in a batch, not with a Quorum Certificate but a proof of inclusion on L1.

At the moment this gossiping of availability and pulling of batches seems to happen whether you need it or not.